examples/fmm/Math3D.cpp

```diff
@@ -0,0 +1,179 @@
+/**
+ * @file Math3D.cpp
+ * @date 23.09.19
+ * @author Joachim Marin
+ */
+
+#include "Math3D.h"
+
+std::vector<double> Math3D::factorialValue;
+std::vector<double> Math3D::doubleFactorialValue;
+std::vector<std::vector<double>> Math3D::getAValue;
+std::vector<Complex> Math3D::imaginaryPower;
+bool Math3D::initialized = false;
+
+std::array<double, 3> Math3D::toSpherical(const std::array<double, 3> &cartesian) {
+  double x = cartesian[0];
+  double y = cartesian[1];
+  double z = cartesian[2];
+  double theta;
+
+  double rho = std::sqrt(x * x + y * y + z * z);
+  double phi = std::atan2(y, x);
+  if (rho > 0) {
+    theta = std::acos(z / rho);
+  } else {
+    theta = 0;
+  }
+
+  if (__isnan(phi)) {
+    std::cerr << "phi is nan" << std::endl;
+  }
+
+  if (__isnan(theta)) {
+    std::cerr << "theta is nan" << std::endl;
+  }
+
+  return std::array<double, 3>({rho, theta /*alpha*/, phi /*beta*/});
+}
+
+// Recurrence relation from https://www.ncbi.nlm.nih.gov/pmc/articles/PMC2634295/#APP1
+double associatedLegendrePolynomialRec(int m, int n, double x) {
+  double ret = 1.0;
+  if (n == m) {
+    // special case n=m=0
+    if (n == 0) {
+      return ret;
+    }
+
+    if (n % 2 == 1) {
+      ret = -1;
+    }
+    ret *= Math3D::doubleFactorial(2 * m - 1);
+    ret *= std::pow(1.0 - x * x, 0.5 * m);
+    return ret;
+  }
+  if (n == m + 1) {
+    return x * (2 * m + 1) * associatedLegendrePolynomialRec(m, m, x);
+  }
+
+  ret = x * (2 * n - 1) * associatedLegendrePolynomialRec(m, n - 1, x) -
+        (n + m - 1) * associatedLegendrePolynomialRec(m, n - 2, x);
+  ret /= (n - m);
+  return ret;
+}
```

Comment:
C++17 implements this in `cmath`:
[https://en.cppreference.com/w/cpp/numeric/special_math/assoc_legendre](https://en.cppreference.com/w/cpp/numeric/special_math/assoc_legendre)
I don't know if the current version of the standard library used in Jenkins already implements this.

Reply:
according to:
https://en.cppreference.com/w/cpp/compiler_support#cpp17
(Mathematical special functions)
it should be there since libstdc++7 and thus also on the jenkins nodes.